src/parser/jobs/blm/modules/RotationWatchdog.tsx

```diff
@@ -0,0 +1,563 @@
+			this.currentRotation.errorCode = ERROR_CODES.DROPPED_ENOCHIAN
+		}
+
+		// Retrieve the GaugeState from the event
+		this.currentGaugeState.astralFire = event.astralFire
+		this.currentGaugeState.umbralIce = event.umbralIce
+		this.currentGaugeState.umbralHearts = event.umbralHearts
+		this.currentGaugeState.enochian = event.enochian
+
+		// If we're in fire phase, stop processing
+		if (this.currentRotation.inFirePhase) {
+			return
+		}
+
+		// If we're in ice phase, set the current gauge state into the pre-fire cache for later recording
+		// Still need by-value assignment here
+		this.currentRotation.gaugeStateBeforeFire = Object.assign(this.currentRotation.gaugeStateBeforeFire, this.currentGaugeState)
+	}
+
+	// Handle cast events and updated recording data accordingly
+	private onCast(event: CastEvent) {
+		const actionId = event.ability.guid
+
+		if (!this.primaryTargetId && event.targetID) {
+			this.primaryTargetId = event.targetID
+		}
+
+		// If this action is signifies the beginning of a new cycle, unless this is the first
+		// cast of the log, stop the current cycle, and begin a new one. If Transposing from ice
+		// to fire, keep this cycle going
+		if (CYCLE_ENDPOINTS.includes(actionId) && !this.firstEvent &&
+			!(actionId === ACTIONS.TRANSPOSE.id && this.currentGaugeState.umbralIce > 0)) {
+			this.startRecording(event)
+		}
+		// Note that we've recorded our first cast now
+		if (this.firstEvent) { this.firstEvent = false }
+
+		// Add actions other than auto-attacks to the rotation cast list
+		const action = getDataBy(ACTIONS, 'id', actionId) as TODO
+		if (action && !action.autoAttack) {
+			this.currentRotation.casts.push(event)
+
+			// If this is manafont, note that we used it so we don't have to cast.filter(...).length to find out
+			if (actionId === ACTIONS.MANAFONT.id) {
+				this.currentRotation.hasManafont = true
+			}
+			// Keep track of total thunder casts so we can include that in the thunder uptime checklist item
+			if (actionId === ACTIONS.THUNDER_III.id && event.targetID === this.primaryTargetId) {
+				this.thunder3Casts++
+			}
+		}
+	}
+
+	// Get the uptime percentage for the Thunder status defbuff
+	private getThunderUptime() {
+		const statusTime = this.enemies.getStatusUptime(STATUSES.THUNDER_III.id)
+		const uptime = this.parser.fightDuration - this.invuln.getInvulnerableUptime()
+
+		return (statusTime / uptime) * 100
+	}
+
+	// Finish this parse and add the suggestions and checklist items
+	private onComplete() {
+		this.stopRecording(undefined)
+		// Suggestion for skipping B4 on rotations that are cut short by the end of the parse or downtime
+		if (this.missedF4s) {
+			this.suggestions.add(new Suggestion({
+				icon: ACTIONS.FIRE_IV.icon,
+				content: <Trans id="blm.rotation-watchdog.suggestions.missed-f4s.content">
+					You lost at least  one <ActionLink {...ACTIONS.FIRE_IV}/> by not skipping <ActionLink {...ACTIONS.BLIZZARD_IV}/> in the Umbral Ice phase before the fight finished.
```

Comment:
```suggestion
					You lost at least one <ActionLink {...ACTIONS.FIRE_IV}/> by not skipping <ActionLink {...ACTIONS.BLIZZARD_IV}/> in the Umbral Ice phase before the fight finished.
```

Reply:
dafeq you doing there space... will fix that